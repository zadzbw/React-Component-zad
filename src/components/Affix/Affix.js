/**
 * Created by zad on 16/12/9.
 */
import React, {PropTypes, Component} from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import _omit from 'lodash/omit';
import _isEqual from 'lodash/isEqual';
import _noop from 'lodash/noop';
import getScroll from '../../utils/getScroll';

import './Affix.less';

function getTargetRect(target) {
  return target === window ? {top: 0, left: 0, bottom: 0} : target.getBoundingClientRect();
}

function getOffset(element, target) {
  const elementRect = element.getBoundingClientRect();
  const targetRect = getTargetRect(target);

  const scrollTop = getScroll(target).top;
  const scrollLeft = getScroll(target).left;

  const clientTop = document.body.clientTop || 0;
  const clientLeft = document.body.clientLeft || 0;

  return {
    top: elementRect.top - targetRect.top + scrollTop - clientTop,
    left: elementRect.left - targetRect.left + scrollLeft - clientLeft
  };
}

export default class Affix extends Component {
  constructor(props) {
    super(props);
    this.state = {
      affixStyle: null,
      affixWrapperStyle: null,
    };
    this._setAffixStyle = this._setAffixStyle.bind(this);
    this._setAffixWrapperStyle = this._setAffixWrapperStyle.bind(this);
    this._updatePosition = this._updatePosition.bind(this);
    this._setEventListener = this._setEventListener.bind(this);
    this._clearEventListener = this._clearEventListener.bind(this);
  }

  static propTypes = {
    offsetTop: PropTypes.number,
    offsetBottom: PropTypes.number,
    target: PropTypes.func,
    onChange: PropTypes.func
  };

  static defaultProps = {
    target: () => window,
    onChange: () => undefined
  };

  _setAffixStyle(e, newAffixStyle) {
    const {onChange = _noop, target} = this.props;
    const prevAffixStyle = this.state.affixStyle;
    const isWindow = target() === window;

    if (e.type === 'scroll' && prevAffixStyle && newAffixStyle && isWindow) {
      return false;
    }

    if (_isEqual(prevAffixStyle, newAffixStyle)) {
      return false;
    }

    this.setState({affixStyle: newAffixStyle}, () => {
      if ((prevAffixStyle && !newAffixStyle) || (!prevAffixStyle && newAffixStyle)) {
        onChange(!!this.state.affixStyle);
      }
    });
  }

  _setAffixWrapperStyle(e, newAffixWrapperStyle) {
    const prevAffixWrapperStyle = this.state.affixWrapperStyle;
    if (e.type === 'resize') {
      return false;
    }

    if (_isEqual(prevAffixWrapperStyle, newAffixWrapperStyle)) {
      return false;
    }
    this.setState({
      affixWrapperStyle: newAffixWrapperStyle
    });
  }

  _updatePosition(e) {
    let {offsetTop, offsetBottom, target} = this.props;
    const targetNode = target();

    const scrollTop = getScroll(targetNode).top;
    const affixNode = ReactDOM.findDOMNode(this);
    const elementOffset = getOffset(affixNode, targetNode);
    const elemSize = {
      width: this.refs.fixedNode.offsetWidth,
      height: this.refs.fixedNode.offsetHeight,
    };

    const offsetMode = {
      top: false,
      bottom: false,
    };
    // 默认使 offsetTop = 0
    if (typeof offsetTop !== 'number' && typeof offsetBottom !== 'number') {
      offsetTop = 0;
      offsetMode.top = true;
    } else {
      offsetMode.top = typeof offsetTop === 'number';
      offsetMode.bottom = typeof offsetBottom === 'number';
    }

    const targetRect = getTargetRect(targetNode);
    const targetInnerHeight = targetNode === window ? targetNode.innerHeight : targetNode.clientHeight;
    if (offsetMode.top && scrollTop > elementOffset.top - offsetTop) {
      // top
      this._setAffixStyle(e, {
        position: 'fixed',
        top: targetRect.top + offsetTop,
        left: targetRect.left + elementOffset.left,
        width: affixNode.offsetWidth,
      });
      this._setAffixWrapperStyle(e, {
        width: elemSize.width,
        height: elemSize.height,
      });
    } else if (offsetMode.bottom && scrollTop < elementOffset.top + elemSize.height + offsetBottom - targetInnerHeight) {
      // bottom
      const targetBottomOffset = targetNode === window ? 0 : (window.innerHeight - targetRect.bottom);
      this._setAffixStyle(e, {
        position: 'fixed',
        bottom: targetBottomOffset + offsetBottom,
        left: targetRect.left + elementOffset.left,
        width: affixNode.offsetWidth,
      });
      this._setAffixWrapperStyle(e, {
        width: elemSize.width,
        height: elemSize.height,
      });
    } else {
      this._setAffixStyle(e, null);
      this._setAffixWrapperStyle(e, null);
    }
  }

  _setEventListener(getTarget) {
    const target = getTarget();
    if (!target) {
      return false;
    }
    target.addEventListener('scroll', this._updatePosition, false);
    target.addEventListener('resize', this._updatePosition, false);
  }

  _clearEventListener(getTarget) {
    const target = getTarget();
    target.removeEventListener('scroll', this._updatePosition, false);
    target.removeEventListener('resize', this._updatePosition, false);
  }

  componentDidMount() {
    const {target} = this.props;
    // next event loop
    this.timeout = setTimeout(() => {
      this._setEventListener(target);
    }, 0);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.target !== this.props.target) {
      this._clearEventListener(this.props.target);
      this._setEventListener(nextProps.target);
      this._updatePosition({});
    }
  }

  componentWillUnmount() {
    const {target} = this.props;
    this._clearEventListener(target);
    clearTimeout(this.timeout);
  }

  render() {
    const affixClass = classNames({
      'zad-affix': this.state.affixStyle
    });
    const props = _omit(this.props, ['offsetTop', 'offsetBottom', 'target']);
    return (
      <div {...props} style={this.state.affixWrapperStyle}>
        <div className={affixClass} ref="fixedNode" style={this.state.affixStyle}>
          {this.props.children}
        </div>
      </div>
    );
  }
}
