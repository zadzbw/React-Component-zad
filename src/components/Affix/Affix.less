@import "../../style/variables";

.zad-affix {
  display: inline-block;
  z-index: @z-index-affix;
}
