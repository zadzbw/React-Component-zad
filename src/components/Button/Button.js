import React, {PropTypes, Component}from 'react';
import classNames from 'classnames';
import oneOf from '../../utils/oneOf';

import './Button.less';

const _type = {
	type: ['primary', 'ghost', 'dashed'],
	size: ['small', 'large'],
	shape: ['circle', 'circle-o']
};
const btnPrefix = 'zad-btn';

export default class Button extends Component {
	constructor(props) {
		super(props);
	}

	static propTypes = {
		type: PropTypes.string,
		size: PropTypes.string,
		shape: PropTypes.string,
		onClick: PropTypes.func
	};

	static defaultProps = {
		onClick: () => undefined
	};


	render() {
		const {type, size, shape, className, children, ...props} = this.props;

		const sizeClass = ({
				large: 'lg',
				small: 'sm',
			})[size] || '';

		const btnClass = classNames(btnPrefix, className, {
			[`${btnPrefix}-${type}`]: type && oneOf(_type.type, type),
			[`${btnPrefix}-${sizeClass}`]: size && oneOf(_type.size, size),
			[`${btnPrefix}-${shape}`]: shape && oneOf(_type.shape, shape)
		});

		return (
			<button
				{...props}
				type="button"
				className={btnClass}
			>
				<span>{children}</span>
			</button>
		);
	}
}
