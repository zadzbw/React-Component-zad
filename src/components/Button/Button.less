@import "../../style/variables";
@import "./ButtonMixin";

.zad-btn {
  .btn();
  .btn-default();

  > .fa + span, > span + .fa, > .fa + .fa {
    margin-left: 0.5em;
  }

  &-primary {
    .btn-primary();
  }

  &-ghost {
    .btn-ghost();
  }

  &-dashed {
    .btn-dashed();
  }

  &-circle {
    .btn-circle();
  }

  &-clicked:before {
    content: '';
    position: absolute;
    border-radius: inherit;
    border: 0 solid @primary-color;
    animation: buttonClick @animation-duration-slow ease-out forwards;
  }
}

@keyframes buttonClick {
  from {
    opacity: 1;
    top: -1px;
    left: -1px;
    bottom: -1px;
    right: -1px;
    border-width: 0;
  }

  to {
    opacity: 0;
    top: -6px;
    left: -6px;
    bottom: -6px;
    right: -6px;
    border-width: 6px;
  }
}