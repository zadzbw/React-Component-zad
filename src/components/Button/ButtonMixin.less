@import "../../style/variables";

.btn-size(@padding;@font-size;@border-radius) {
  padding: @padding;
  font-size: @font-size;
  border-radius: @border-radius;
}

.btn-color(@color;@background;@border) {
  color: @color;
  background-color: @background;
  border-color: @border;
}

.btn-disabled() {
  &[disabled] {
    cursor: not-allowed;
    > * {
      pointer-events: none;
    }

    &, &:hover, &:focus, &:active {
      .btn-color(@btn-disable-color; @btn-disable-bg; @btn-disable-border);
    }
  }
}

.btn-variant-primary(@color;@background) {
  .btn-color(@color; @background; @background);

  &:hover, &:focus {
    .btn-color(@color; tint(@background, 20%); tint(@background, 20%));
  }

  &:active {
    .btn-color(@color; shade(@background, 5%); shade(@background, 5%));
  }

  .btn-disabled();
}

.btn-variant-others(@color;@background;@border) {
  .btn-color(@color; @background; @border);

  &:hover, &:focus {
    .btn-color(tint(@primary-color, 20%); @background; tint(@primary-color, 20%));
  }

  &:active {
    .btn-color(shade(@primary-color, 5%); @background; shade(@primary-color, 5%));
  }

  .btn-disabled();
}

.btn() {
  position: relative;
  display: inline-block;
  text-align: center;
  font-weight: @btn-font-weight;
  line-height: @line-height-base;
  transition: all @animation-duration-base @ease-in-out;
  margin-bottom: 0;
  border: 1px solid transparent;
  touch-action: manipulation;
  cursor: pointer;
  background-image: none;
  white-space: nowrap;
  user-select: none;

  &, &:active, &:focus {
    outline: 0;
  }

  &:not([disabled]):hover {
    text-decoration: none;
  }

  &:not([disabled]):active {
    outline: 0;
    transition: none;
  }

  .btn-size(@btn-padding-base; @font-size-base; @btn-border-radius-base);

  &-sm {
    .btn-size(@btn-group-padding-sm; @font-size-sm; @btn-border-radius-sm);
  }

  &-lg {
    .btn-size(@btn-padding-lg; @btn-font-size-lg; @btn-border-radius-base);
  }
}

.btn-default() {
  &:hover, &:focus, &:active, &.active {
    background: #fff;
  }
  .btn-variant-others(@btn-default-color; @btn-default-bg; @btn-default-border);
}

.btn-primary() {
  .btn-variant-primary(@btn-primary-color; @btn-primary-bg);
}

.btn-ghost() {
  .btn-variant-others(@btn-ghost-color, @btn-ghost-bg, @btn-ghost-border);
}

.btn-dashed() {
  .btn-ghost();
  border-style: dashed;
}

.btn-circle() {
  top: 1px; // 清除circle向上偏移的1px
  width: @btn-circle-size-base;
  height: @btn-circle-size-base;
  .btn-size(0; @font-size-base + 2; 50%);

  &.zad-btn-sm {
    top: 0;
    width: @btn-circle-size-sm;
    height: @btn-circle-size-sm;
    .btn-size(0; @font-size-sm + 2; 50%);
  }

  &.zad-btn-lg {
    top: 0;
    width: @btn-circle-size-lg;
    height: @btn-circle-size-lg;
    .btn-size(0; @btn-font-size-lg + 2; 50%);
  }
}
