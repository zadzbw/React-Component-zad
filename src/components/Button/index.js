/**
 * Created by zad on 16/11/30.
 */
import Button from './Button';

export default Button;