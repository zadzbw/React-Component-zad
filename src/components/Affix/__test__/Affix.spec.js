/**
 * Created by zad on 16/12/9.
 */
import React from 'react';
import {mount} from 'enzyme';
import {mountToJson} from 'enzyme-to-json';
import Affix from '../../Affix';
import Button from '../../Button';

describe('Affix test', () => {
  it('normal', () => {
    const wrapper = mount(
      <Affix offsetTop={20}>
        <Button type="primary">Affix top</Button>
      </Affix>
    );
    expect(mountToJson(wrapper)).toMatchSnapshot();
  });
});
