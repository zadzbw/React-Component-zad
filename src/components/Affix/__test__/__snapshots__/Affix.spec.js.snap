exports[`Affix test normal 1`] = `
<Affix
  offsetTop={20}
  onChange={[Function]}
  target={[Function]}>
  <div
    onChange={[Function]}
    style={null}>
    <div
      className=""
      style={null}>
      <Button
        onClick={[Function]}
        type="primary">
        <button
          className="zad-btn zad-btn-primary"
          onClick={[Function]}
          onMouseUp={[Function]}
          type="button">
          <span>
            Affix top
          </span>
        </button>
      </Button>
    </div>
  </div>
</Affix>
`;
